File: 13. Rekursi Lanjut/C. Permutasi Zig-Zag/c_permutasi_zig_zag.go
package main

import (
	"fmt"
)

var save []int
var isVisited []bool
var N int

func write(depth int) {
	if depth > N {
		isZigZag := true
		if N >= 3 {
			for i := 2; i < N; i++ {
				h := save[i] > save[i-1] && save[i] > save[i+1]
				l := save[i] < save[i-1] && save[i] < save[i+1]
				if !(h || l) {
					isZigZag = false
					break
				}
			}
		}
		if N < 3 || isZigZag {
			for i := 1; i <= N; i++ {
				fmt.Print(save[i])
			}
			fmt.Println()
		}

	} else {
		for i := 1; i <= N; i++ {
			if !isVisited[i] {
				isVisited[i] = true
				save[depth] = i
				write(depth + 1)
				isVisited[i] = false

			}
		}
	}
}
func main() {
	save = make([]int, 17)
	isVisited = make([]bool, 17)
	fmt.Scan(&N)
	write(1)
}

File: 13. Rekursi Lanjut/C. Permutasi Zig-Zag/c_permutasi_zig_zag.java
import java.util.Scanner;

public class Solution {
    public static int[] save = new int[17];
    public static boolean[] isVisited = new boolean[17];
    public static int N;

    public static void write(int depth) {
        if (depth > N) {
            boolean isZigZag = true;
            if (N >= 3) {
                for (int i = 2; i < N; i++) {
                    boolean h = (save[i] > save[i - 1]) && (save[i] > save[i + 1]);
                    boolean l = (save[i] < save[i - 1]) && (save[i] < save[i + 1]);
                    if (!(h || l)) {
                        isZigZag = false;
                        break;
                    }
                }
            }
            if (N < 3 || isZigZag) {
                for (int i = 1; i <= N; i++) {
                    System.out.print(save[i]);
                }
                System.out.println();
            }
        } else {
            for (int i = 1; i <= N; i++) {
                if (!isVisited[i]) {
                    isVisited[i] = true;
                    save[depth] = i;
                    write(depth + 1);
                    isVisited[i] = false;
                }
            }
        }
    }

    public static void main(String[] args) {
        Scanner scan = new Scanner(System.in);
        N = scan.nextInt();
        scan.close();
        write(1);
    }
}
